recipe/2024-06-private-bicep-registries.md

```diff
@@ -0,0 +1,269 @@
+
+### Implementation Details
+#### Portable Resources / Recipes RP (if applicable)
+***Bicep Driver***
+
+- Implement FindSecretIDs() function in bicep driver, to retrieve the secret store resource IDs associated with private bicep registries.
+    ```
+    // FindSecretIDs is used to retrieve a map of secretStoreIDs and corresponding secret keys.
+    // associated with the given environment configuration and environment definition.
+    func (d *bicepDriver) FindSecretIDs(ctx context.Context, envConfig recipes.Configuration, definition recipes.EnvironmentDefinition) (secretStoreIDResourceKeys map[string][]string, err error) {
+        ...
+    }
+    ```
+- Update Bicep driver apis i.e execute, delete and getMetadata to use oras auth package for private registry authentication.
+
+### Error Handling
+
+- Authorization failure error:
+
+  This error can occur if users provide incorrect credential information or omit the Bicep recipe configuration for their private registry, we can add a new error code for this scenario `BicepRegistryAuthFailed`
+  
+  Incorrect credential information:
+  ```
+  NewRecipeError("BicepRegistryAuthFailed", fmt.Sprintf("failed to authenticate the bicep registry %s: %s", <private-oras-registry-name>, <error returned by the oras client>))
+  ```
+  Missing Bicep Recipe Config:
+  ```
+  NewRecipeError("BicepRegistryAuthFailed", fmt.Sprintf("missing recipe config for registry %s: %s", <private-oras-registry-name>, <error returned by the oras client>))
+  ```
+
+## Test plan
+
+#### Unit Tests
+-   Update environment conversion unit tests to validate recipeConfig property.
+-   Update environment controller unit tests to add recipe config.
+-   Adding new unit tests in bicep driver validating recipe config changes and retrieving secrets.
+
+#### Functional tests
+- Add e2e test to verify recipe deployment using a bicep stored in a private bicep registry.
+    - publish a recipe to private ghcr.io/radius-project/private-recipes/<recipe> 
+    - Deploy the recipe as part of the functional test using github app token to authenticate ghcr. 
+
+## Security
+With this design we enable username-password based authentication for OCI compliant registries, we let the users manage secrets. For secret rotation users need to re deploy the `Applications.Core/secretStores` resource with updated credentials.
+
+## Development plan
+
+- Task 1:  
+    - Adding a new property bicepConfig in recipeConfig to environment and making necessary changes to typespec, datamodel and conversions.
+    - Updating unit tests.
+- Task 2:
+    - Adding changes to bicep driver to retrieve secrets.
+    - Adding changes to bicep driver to use oras auth package to authenticate private bicep registries
+    - Update/Add unit tests
+- Task 3:
+    - Adding cli changes for `rad bicep publish` to support private bicep registries.
+    - Update the cli unit tests.
+- Task 3:
+    - Manual Validation and adding e2e tests to verify using private bicep registries
```

Comment:
also need to update and add developer documentation.

Reply:
added it